File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Node{
public:
    int stops;
    int flight;
    int price;
    Node(int stops,int flight,int price){
        this->stops = stops;
        this->flight = flight;
        this->price = price;
    }
};
class Edge{
public:
    int neiFlight;
    int price;
    Edge(int neiFlight,int price){
        this->neiFlight = neiFlight;
        this->price = price;
    }
};
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        unordered_map<int,vector<Edge> > flightGraph;
        int length = flights.size();
        for(int index=0;index<length;++index){
            int source = flights[index][0];
            int dest = flights[index][1];
            int price = flights[index][2];
            flightGraph[source].push_back(Edge(dest,price));
        }
        vector<int> priceArray(n,INT_MAX);
        priceArray[src] = 0;
        queue<Node> flightQueue;
        flightQueue.push(Node(0,src,0));

        while(!flightQueue.empty()){
            int currentQueueSize = flightQueue.size();
            while(currentQueueSize > 0){
                Node currentFlightNode = flightQueue.front();
                flightQueue.pop();
                vector<Edge> neiFlights;
                if(flightGraph.find(currentFlightNode.flight) != flightGraph.end()){
                    neiFlights = flightGraph[currentFlightNode.flight];
                }
                cout<<neiFlights.size()<<endl;
                for(int index=0;index<neiFlights.size();++index){
                    int flightName = neiFlights[index].neiFlight;
                    int flightPrice = neiFlights[index].price;
                    if(currentFlightNode.price + flightPrice < priceArray[flightName] && currentFlightNode.stops + 1 <= k + 1){
                        priceArray[flightName] = currentFlightNode.price + flightPrice;
                        flightQueue.push(Node(currentFlightNode.stops + 1,flightName,priceArray[flightName]));
                    }
                }

                currentQueueSize--;
            }
        }
        if(priceArray[dst] == INT_MAX){
            return -1;
        }
        return priceArray[dst];
    }
};

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.java
class Solution {
    class Node {
        public int destination;
        public int price;

        public Node(int destination, int price) {
            this.destination = destination;
            this.price = price;
        }
    }

    class FlightTracker {
        public int flight;
        public int stops;
        
        public int cost;
        public FlightTracker(int flight, int stops, int cost) {
            this.flight = flight;
            this.stops = stops;
            this.cost = cost;
        }
    }
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        //build the graph
        HashMap<Integer,List<Node>> graph = new HashMap<>();
        for(int[] flight : flights) {
            int sourceFlight = flight[0];
            int destFlight = flight[1];
            int price = flight[2];
            graph.putIfAbsent(sourceFlight, new ArrayList<>());
            List<Node> list = graph.get(sourceFlight);
            list.add(new Node(destFlight, price));
            graph.put(sourceFlight, list);
        }

        PriorityQueue<FlightTracker> q = new PriorityQueue<>((a,b) -> Integer.compare(b.cost,a.cost));
        int[] distance = new int[n];
        Arrays.fill(distance, Integer.MAX_VALUE);
        distance[src] = 0;
        q.add(new FlightTracker(src, 0,0));
        while(!q.isEmpty()) {
            int size = q.size();
            for(int index=0;index<size;++index) {
                FlightTracker currentFlight = q.poll();
                // if(currentFlight.flight == dst) {
                //     return distance[dst];
                // }
                List<Node> adajacentFlights = graph.getOrDefault(currentFlight.flight, new ArrayList<>());
                for(Node node : adajacentFlights) {
                     if(currentFlight.stops <= k && (currentFlight.cost + node.price < distance[node.destination])) {
                        distance[node.destination] = currentFlight.cost + node.price;
                        q.add(new FlightTracker(node.destination, currentFlight.stops + 1,distance[node.destination]));
                    }
                }
            }
        }

        return (distance[dst] == Integer.MAX_VALUE) ? -1 : distance[dst];
    }
}